### DISCUSSION TITLE
Parse middleware schema inputs

### DISCUSSION THREAD

**Discussion by gdaybrice** [next-safe-action/next-safe-action#254] [Q&A]
Is there a way to get `parsedInput` within the `use` middleware? Currently only having access to `clientInput` even if a schema is defined.

**Comment by TheEdoRan** [comment]
Yes, you can await the result of the action execution with the `next` function, and then return the result:

![image](https://github.com/user-attachments/assets/78cb71fa-eea4-4858-9dfb-70143d43c66b)

**Reply by gdaybrice** [reply]
Thanks @TheEdoRan - any way to get the parsed inputs as arguments to the middleware? I need to check if a user belongs to an organisation before it runs the action.

At the moment I need to get the clientInput and manually parse with a zod schema which seems a bit redundant.

I'm migrating from zsa and that something I found useful, especially if a parent schema is attached to the client. 

It would be great if schemas are automatically parsed and extended in order.

**Reply by TheEdoRan** [reply]
> any way to get the parsed inputs as arguments to the middleware?
> I need to check if a user belongs to an organisation before it runs the action.

In next-safe-action, middleware functions are called **before** the action server code function. So, if you define a middleware that throws an error on a specific check, when you're calling that action from the client, the code never reaches input validation and the action server code body. If instead you `await` the `next` function, it recursively calls all the middleware defined after that one, and then the action server code function, at the end of the chain.

> It would be great if schemas are automatically parsed and extended in order.

[You can already extend the schema](https://next-safe-action.dev/docs/define-actions/extend-previous-schemas), but the input is parsed and validated when you're calling the action, so inside `action` method. `schema` is just used to pass a validation schema to that particular safe action client. Otherwise, It would be impossible to guarantee that you passed the right input to the action, since you can only do it when you're calling it from the client.
